### DISCUSSION TITLE
'Fix on Save' Only Works with Explicit Rule Options in VS Code (Configuration Issue? Bug?)

### DISCUSSION THREAD

**Discussion by oschade** [biomejs/biome#3543] [Q&A] [answered]
Hi,

I'm experiencing an issue that might be a bug, but I suspect it's related to configuration. In my VS Code, the "fix-on-save" option for rules like "useSelfClosingElements" and "useTemplate" only works if I explicitly set {"fix": "safe"} in the linter rule options:

```json
{
	"$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
	"organizeImports": {
		"enabled": true
	},
	"linter": {
		"enabled": true,
		"rules": {
			"recommended": true,
			"style": {
				"noNonNullAssertion": "warn",
				"useNamingConvention": {
					"level": "warn",
					"options": {
						"enumMemberCase": "CONSTANT_CASE",
						"strictCase": false
					}
				},
				"useSelfClosingElements": {
					"level": "warn",
					"fix": "safe"
				},
				"useTemplate": {
					"level": "warn",
					"fix": "safe"
				}
			},
			"suspicious": {
				"noExplicitAny": "off"
			},
			"complexity": {
				"noForEach": "off"
			},
			"correctness": {
				"useExhaustiveDependencies": "off",
				"noUnusedImports": "warn",
				"noUnusedVariables": "warn",
				"noInvalidUseBeforeDeclaration": "warn"
			}
		}
	},
	"formatter": {
		"formatWithErrors": true
	},
	"javascript": {
		"formatter": {
			"arrowParentheses": "always",
			"jsxQuoteStyle": "double",
			"semicolons": "asNeeded",
			"trailingCommas": "es5",
			"quoteStyle": "single",
			"indentStyle": "space"
		}
	},
	"json": {
		"formatter": {
			"enabled": true
		},
		"parser": {
			"allowComments": true,
			"allowTrailingCommas": true
		}
	}
}

```

Is this the intended behavior? Can I set all recommended rules to {"fix": "safe"} globally, or do I need to specify this option for each rule individually?

Thank you for your assistance!

**Comment by Conaclos** [comment, marked as answer]
Biome has the concept of [safe fixes](https://biomejs.dev/linter/#safe-fixes) and [unsafe fixes](https://biomejs.dev/linter/#unsafe-fixes). Fix-on-save applies only safe fixes. As you figured out, you can change the safety of a fix. By marking it as `safe` you allows them to be fixed-on-save.

> Is this the intended behavior? Can I set all recommended rules to {"fix": "safe"} globally, or do I need to specify this option for each rule individually?

It is not possible to mark all fixes as safe for now.
Thus, you have to set them for every rule you want to fix-on-save.
This avoids bad surprises and requires you to think carefully each time you want to apply automatically some fixes.

Fixes of some rules could be reworked

**Reply by oschade** [reply]
Ah, okay, I read the setting as "fix it, when it's safe" instead of "treat the fix as safe". Thank you!

**Reply by Conaclos** [reply]
Which settings? Perhaps we could improve the wording.

**Reply by oschade** [reply]
I mean this:

```
"useSelfClosingElements": {
    "level": "warn",
    "fix": "safe"
},
```

Maybe you could call this "fix-level" or something.

**Reply by Conaclos** [reply]
Thanks for the feedback.
